### ISSUE TITLE
Trainging with peft costs more time than trainging the whole model

### ISSUE THREAD

**Issue by Nidhogg-lyz** [huggingface/peft#960]
### System Info

Hi, thanks for your contributions! Now I'm learning to use peft for finetuning. My code runs on google colab T4 with peft version 0.5.0.

### Information

- [ ] The official example scripts
- [X] My own modified scripts

### Tasks

- [ ] An officially supported task in the `examples` folder
- [X] My own task or dataset (give details below)

### Reproduction

I'm learning how to use peft and I use lora to finetune ResNet18 on CIFAR10. 
The original code in colab is [here](https://colab.research.google.com/drive/1AeGQoTJcMrRakXicBKUgrH0ZCdDpVyoj?usp=sharing)
```
origin_model = resnet18(weights=None, num_classes=10)
origin_model.to(DEVICE)
target_modules = []
available_types = [torch.nn.modules.conv.Conv2d, torch.nn.modules.linear.Linear]
for n, m in origin_model.named_modules():
  if type(m) in available_types:
    target_modules.append(n)
target_modules.remove('fc')

config = LoraConfig(
    r=16,
    lora_alpha=16,
    lora_dropout=0.1,
    bias="none",
    target_modules=target_modules,
    modules_to_save=["fc"],
)
peft_model = get_peft_model(origin_model, config).to(DEVICE)
peft_model.print_trainable_parameters()
```
The trainable parameters and ratio is _trainable params: 585,028 || all params: 11,766,340 || trainable%: 4.9720473826185545_.
Below is my training loop:
```
def my_train(net, train_loader):
    epochs = 10
    criterion = nn.CrossEntropyLoss()
    trainable_para = []
    for p in net.parameters():
      if p.requires_grad:
        trainable_para.append(p)
    print("num of trainable parameters: ", sum(p.numel() for p in trainable_para if p.requires_grad))
    optimizer = optim.Adam(trainable_para, lr=0.001)
    scheduler = torch.optim.lr_scheduler.CosineAnnealingLR(optimizer, T_max=epochs)
    net.train()
    for _ in range(epochs):
        for inputs, targets in tqdm(train_loader):
            inputs, targets = inputs.to(DEVICE), targets.to(DEVICE)
            optimizer.zero_grad()
            outputs = net(inputs)
            loss = criterion(outputs, targets)
            loss.backward()
            optimizer.step()
        scheduler.step()
    net.eval()
```

Then I train the peft_model and compare the time cost with training the whole model:
```my_train(peft_model, train_loader)```
```
model = resnet18(weights=None, num_classes=10)
model.to(DEVICE)
my_train(model, train_loader)
```
But even the fastest epoch when training peft_model costs more time than training the whole model...
Is this behavior normal? Or is there something wrong with my code?

### Expected behavior

The time cost of training peft_model should be less than training the whole model.

**Comment by Nidhogg-lyz** [comment]
> Thanks for reporting and providing the code.
> 
> > But even the fastest epoch when training peft_model costs more time than training the whole model...
> 
> I ran your notebook on my machine and could not reproduce this issue (or maybe I misunderstand what you mean). For me, I get basically the same speed with LoRA vs full fine-tuning, so there is no slowdown when using LoRA.
> 
> Note that using LoRA does not automatically make training faster. After all, LoRA adds more parameters to the model during the forward pass, making it a bit slower (depending on model architecture and hyper-params). Of course, the model update should be faster with LoRA, but in sum, there is not necessarily a speed advantage. However: Since LoRA training should reduce the memory footprint, you can try increasing the batch size, which should lead to a speed up. The larger the model, the more you should benefit from this. You can also experiment with higher learning rates when using LoRA, which would reduce the number of epochs required for convergence.
> 
> Something else I observed when running your code is that GPU usage is well below 100%, so there could be room for improvement to optimize GPU utilization. I didn't further investigate this though.

@BenjaminBossan Thanks for the reply!
What I mean is that: for each training epoch, the training of peft_model is slower than training the whole model for one epoch(20s per epoch vs. 17s per epoch). This means if I have enough memory to train the whole model, there is no advantage in time cost when using lora. As you said the peft_model may be a bit slower during the forward pass, but I think the calculation of gradients and parameter updating is much more time-consuming, and due to the much fewer(about 5%) trainable parameters in peft_model, the time cost should be less than training the whole model.
Is the gradient calculation cost between peft_model and origin_model so close that the major time difference comes from the forward pass?
Or is the batch_size too small that the code cannot show the time benefit of peft_model?

**Comment by BenjaminBossan** [comment]
> the training of peft_model is slower than training the whole model for one epoch(20s per epoch vs. 17s per epoch)

As mentioned, for me there was no different, LoRA was even ever so slightly faster, so this seems to be depend on hardware. When it comes to whether the disadvantage in the forward pass is made up for with the advantage in the backwards pass, it is at the end an empirical question and will depend on circumstances. From what I know, there isn't a general observation that applying LoRA, without any further changes, leads to significant speed up in training tie.

I would recommend experimenting with higher batch size and/or learning rate when using LoRA, hopefully this helps accelerating the training for you.

**Comment by BenjaminBossan** [comment]
> Even the PEFT webpage states:

True, I agree that the description should focus more on the memory part, which is the essence of PEFT. When it comes to speed, it's really hard to say, as it depends on so many factors - what PEFT method is used, which model, which hyper-parameters, etc.

> I haven't noticed any significant gains in training speed when using PEFT and LoRA

I can't really add much to what I wrote earlier: You should have more memory available, so higher batch size should work. You should also try increasing the learning rate compared to full fine tuning (see e.g. section 4.6.1 of [this paper](https://arxiv.org/abs/2405.09673)). With higher learning rate, you may also see faster convergence.

And if LoRA allows you to train a bigger model that would otherwise not be possible to train, well, time difference is finite vs infinite ;)
